// ngx-persian/esm5/public_api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of ngx-persian
 */
export { NgxPersianModule } from './lib/ngx-persian.module';
export { PersianLetterService } from './lib/Services/persian-letter.service';
export { PersianNumberService } from './lib/Services/persian-number.service';
export { EnNumPipe } from './lib/Pipes/enNum.pipe';
export { FaNumPipe } from './lib/Pipes/faNum.pipe';
export { IRCurrencyPipe } from './lib/Pipes/IRCurrency.pipe';
export { JdatePipe } from './lib/Pipes/jdate.pipe';
export { NationalCodePipe } from './lib/Pipes/nationalCode.pipe';
export { JDate } from './lib/JDate/jdate';
export { JalaliDateCalculatorService } from './lib/JDate/jalali-date-calculator.service';
export { JalaliDateValidatorService } from './lib/JDate/jalali-date-validator.service';
export { persianLettersValidator, persianNumbersValidator } from './lib/Validators/persian-validators';
export { PersianNumbersDirective, PersianLetterDirective } from './lib/Validators/persian-directives';
export { MobileCodes, MobilePhoneNumberService, operatorsNames } from './lib/Services/mobile-phone-number.service';
export { NationalCodeService } from './lib/Services/national-code.service';
