// ngx-persian/esm5/lib/Pipes/IRCurrency.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import { DecimalPipe } from '@angular/common';
/** @enum {string} */
var IRCurrencies = {
    rial: 'ریال',
    r: 'ریال',
    toman: 'تومان',
    t: 'تومان',
};
export { IRCurrencies };
/**
 * Formats input as a currency value.
 * default type is rial. You can choose on of the following currency types:
 *
 *      r or rial for ریال
 *
 *      to or toman for تومان
 *
 *  This pipe extends DecimalPipe, so as second parameter, you can enter your desired formatting string. Default formatter is: 1.0-0
 */
var IRCurrencyPipe = /** @class */ (function (_super) {
    tslib_1.__extends(IRCurrencyPipe, _super);
    function IRCurrencyPipe() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param value a number of a string only contains digits
     * @param type currency type
     * @param digitInfo decimal pipe formatter
     * @example 1925100 -> 1,925,100 ریال
     */
    /**
     * \@example 1925100 -> 1,925,100 ریال
     * @param {?} value a number of a string only contains digits
     * @param {?=} type currency type
     * @param {?=} digitInfo decimal pipe formatter
     * @return {?}
     */
    IRCurrencyPipe.prototype.transform = /**
     * \@example 1925100 -> 1,925,100 ریال
     * @param {?} value a number of a string only contains digits
     * @param {?=} type currency type
     * @param {?=} digitInfo decimal pipe formatter
     * @return {?}
     */
    function (value, type, digitInfo) {
        if (type === void 0) { type = 'rial'; }
        if (digitInfo === void 0) { digitInfo = '1.0-0'; }
        value = Number(value);
        type = type.toLowerCase();
        if (isNaN(value)) {
            throw new Error(value + " is not a acceptable number");
        }
        return _super.prototype.transform.call(this, value, digitInfo) + ' ' + IRCurrencies[type];
    };
    IRCurrencyPipe.decorators = [
        { type: Pipe, args: [{ name: 'irc' },] }
    ];
    return IRCurrencyPipe;
}(DecimalPipe));
export { IRCurrencyPipe };
