// ngx-persian/esm2015/lib/Pipes/IRCurrency.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { DecimalPipe } from '@angular/common';
/** @enum {string} */
const IRCurrencies = {
    rial: 'ریال',
    r: 'ریال',
    toman: 'تومان',
    t: 'تومان',
};
export { IRCurrencies };
/**
 * Formats input as a currency value.
 * default type is rial. You can choose on of the following currency types:
 *
 *      r or rial for ریال
 *
 *      to or toman for تومان
 *
 *  This pipe extends DecimalPipe, so as second parameter, you can enter your desired formatting string. Default formatter is: 1.0-0
 */
export class IRCurrencyPipe extends DecimalPipe {
    /**
     * \@example 1925100 -> 1,925,100 ریال
     * @param {?} value a number of a string only contains digits
     * @param {?=} type currency type
     * @param {?=} digitInfo decimal pipe formatter
     * @return {?}
     */
    transform(value, type = 'rial', digitInfo = '1.0-0') {
        value = Number(value);
        type = type.toLowerCase();
        if (isNaN(value)) {
            throw new Error(`${value} is not a acceptable number`);
        }
        return super.transform(value, digitInfo) + ' ' + IRCurrencies[type];
    }
}
IRCurrencyPipe.decorators = [
    { type: Pipe, args: [{ name: 'irc' },] }
];
