// ngx-persian/esm2015/lib/Validators/persian-directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { persianLettersValidator, persianNumbersValidator } from './persian-validators';
/**
 * You can add PersianNumbersDirective property to your input tag for using persianNumbersValidator function on its value.
 *
 * [For more information also see persianNumbersValidator]{\@link persianNumbersValidator}
 */
export class PersianNumbersDirective {
    /**
     * Only a wrapper for persianNumbersValidator validator function.
     * @param {?} control
     * @return {?} null for success or ValidationError for invalid inputs
     */
    validate(control) {
        return persianNumbersValidator()(control);
    }
}
PersianNumbersDirective.decorators = [
    { type: Directive, args: [{
                selector: 'PersianNumbersDirective',
                providers: [{ provide: NG_VALIDATORS, useExisting: PersianNumbersDirective }]
            },] }
];
/**
 * You can add PersianLetterDirective property to your input tag for using persianNumbersValidator function on its value
 * For setting options value, you can simply set PersianLetterDirective property equal to the PLOptions object.
 * [For more information also see persianLettersValidator]{\@link persianLettersValidator}
 */
export class PersianLetterDirective {
    /**
     * Only a wrapper for persianLettersValidator validator function.
     * @param {?} control
     * @return {?} null for success or ValidationError for invalid inputs
     */
    validate(control) {
        return persianLettersValidator(this.options)(control);
    }
}
PersianLetterDirective.decorators = [
    { type: Directive, args: [{
                selector: 'persianLettersValidator',
                providers: [{ provide: NG_VALIDATORS, useExisting: PersianLetterDirective }]
            },] }
];
PersianLetterDirective.propDecorators = {
    options: [{ type: Input, args: ['PersianLetterDirective',] }]
};
if (false) {
    /** @type {?} */
    PersianLetterDirective.prototype.options;
}
