// ngx-persian/esm2015/lib/Validators/persian-validators.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PersianNumberService } from '../Services/persian-number.service';
import { PersianLetterService } from '../Services/persian-letter.service';
/**
 * Reactive form validator that checks form control value contains only persian numbers.
 * @param {?=} persianNumberService
 * @return {?}
 */
export function persianNumbersValidator(persianNumberService = new PersianNumberService()) {
    return (/**
     * @param {?} control
     * @return {?}
     */
    (control) => {
        return persianNumberService.isPersian(control.value) ? null : { 'persianNumberValidation': control.value };
    });
}
/**
 * Reactive form validator that checks form control value contains only persian letters and allowed characters in the options parameter.
 * @param {?=} options see documents of isPersian method of the PersianLetterService for more info about options.
 * @param {?=} persianLetterService
 * @return {?}
 */
export function persianLettersValidator(options = {
    persianDigits: false,
    whitespaces: false,
    symbols: false,
    enDigits: false
}, persianLetterService = new PersianLetterService()) {
    return (/**
     * @param {?} control
     * @return {?}
     */
    (control) => {
        return persianLetterService.isPersian(control.value, options) ? null : { 'persianLetterValidation': control.value };
    });
}
