// ngx-persian/esm5/lib/Services/persian-number.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { InvalidServiceInputError } from './InvalidServiceInput.error';
import * as i0 from "@angular/core";
var PersianNumberService = /** @class */ (function () {
    function PersianNumberService() {
    }
    /**
     * Checks if input value contains any persian number or not.
     * @param value
     * @return true if input value contains any persian number. false otherwise.
     */
    /**
     * Checks if input value contains any persian number or not.
     * @param {?} value
     * @return {?} true if input value contains any persian number. false otherwise.
     */
    PersianNumberService.prototype.containsPersian = /**
     * Checks if input value contains any persian number or not.
     * @param {?} value
     * @return {?} true if input value contains any persian number. false otherwise.
     */
    function (value) {
        if (!value) {
            return false;
        }
        // language=JSRegexp
        /** @type {?} */
        var persianRegex = new RegExp("[" + PersianNumberService.persianNumberPattern + "]");
        return persianRegex.test(value);
    };
    /**
     * splits input value by persian letters and then returns length of matched array.
     * @param value
     * @return 0 if input value has no persian letter, numbers of persian letters in the input value otherwise.
     */
    /**
     * splits input value by persian letters and then returns length of matched array.
     * @private
     * @param {?} value
     * @return {?} 0 if input value has no persian letter, numbers of persian letters in the input value otherwise.
     */
    PersianNumberService._getMatchedPatternLength = /**
     * splits input value by persian letters and then returns length of matched array.
     * @private
     * @param {?} value
     * @return {?} 0 if input value has no persian letter, numbers of persian letters in the input value otherwise.
     */
    function (value) {
        // language=JSRegexp
        /** @type {?} */
        var matchResult = value.match(new RegExp("[" + PersianNumberService.persianNumberPattern + "]", 'g'));
        return matchResult ? matchResult.length : 0;
    };
    /**
     * Checks if input value contains only persian letters.
     * @param value
     */
    /**
     * Checks if input value contains only persian letters.
     * @param {?} value
     * @return {?}
     */
    PersianNumberService.prototype.isPersian = /**
     * Checks if input value contains only persian letters.
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (!value) {
            return false;
        }
        return value.length === PersianNumberService._getMatchedPatternLength(value);
    };
    /**
     * Converts arabic numbers to the persian ones.
     * @param value
     */
    /**
     * Converts arabic numbers to the persian ones.
     * @param {?} value
     * @return {?}
     */
    PersianNumberService.prototype.arabicToPersian = /**
     * Converts arabic numbers to the persian ones.
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value.replace(/٤/g, PersianNumberService.persianNumbersTable[4])
            .replace(/٥/g, PersianNumberService.persianNumbersTable[5])
            .replace(/٦/g, PersianNumberService.persianNumbersTable[6]);
    };
    /**
     * Converts all arabic and english numbers to the persian numbers.
     * @param value
     */
    /**
     * Converts all arabic and english numbers to the persian numbers.
     * @param {?} value
     * @return {?}
     */
    PersianNumberService.prototype.toPersian = /**
     * Converts all arabic and english numbers to the persian numbers.
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value === undefined || value === null) {
            throw new InvalidServiceInputError();
        }
        value = value.toString();
        value = this.arabicToPersian(value);
        /** @type {?} */
        var regex;
        for (var i = 0; i < PersianNumberService.persianNumbersTable.length; i++) {
            // language=JSRegexp
            regex = new RegExp("[" + PersianNumberService.englishNumbersTable[i] + "]", 'g');
            value = value.replace(regex, PersianNumberService.persianNumbersTable[i]);
        }
        return value;
    };
    /**
     * Convert persian numbers in input value to the english numbers.
     * @param value
     */
    /**
     * Convert persian numbers in input value to the english numbers.
     * @param {?} value
     * @return {?}
     */
    PersianNumberService.prototype.toEnglish = /**
     * Convert persian numbers in input value to the english numbers.
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value === undefined || value === null) {
            throw new InvalidServiceInputError();
        }
        /** @type {?} */
        var regex;
        for (var i = 0; i < PersianNumberService.englishNumbersTable.length; i++) {
            // language=JSRegexp
            regex = new RegExp("[" + PersianNumberService.persianNumbersTable[i] + "]", 'g');
            value = value.replace(regex, PersianNumberService.englishNumbersTable[i]);
        }
        return value;
    };
    PersianNumberService.persianNumberPattern = '\u06F0-\u06F9';
    PersianNumberService.persianNumbersTable = ['\u06F0', '\u06F1', '\u06F2', '\u06F3', '\u06F4', '\u06F5', '\u06F6', '\u06F7', '\u06F8', '\u06F9'];
    PersianNumberService.englishNumbersTable = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
    PersianNumberService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ PersianNumberService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function PersianNumberService_Factory() { return new PersianNumberService(); }, token: PersianNumberService, providedIn: "root" });
    return PersianNumberService;
}());
export { PersianNumberService };
if (false) {
    /** @type {?} */
    PersianNumberService.persianNumberPattern;
    /**
     * @type {?}
     * @private
     */
    PersianNumberService.persianNumbersTable;
    /**
     * @type {?}
     * @private
     */
    PersianNumberService.englishNumbersTable;
}
